Integrate Circom into Bellperson for Coprocessors

winston-h-zhang opened the pull request:
This PR depends on an extension of Nova-Scotia and ArkCircom, here: https://github.com/huitseeker/Nova-Scotia/tree/winston-main

This (draft) PR adds some API to create coprocessors from circom. In particular:
- We need to keep some Circom metadata needed to instantiate bellperson gadgets. This data is stored in `.lurk/circom-gadgets/*` and the intention is for users to manipulate this through the CLI. There's a very rough draft of what that should look like in this PR.
- For the discussion [here](https://zulip.lurk-lab.com/#narrow/stream/45-crypto/topic/Circom.20conversion/near/97007), as I wrote out the API, I found that my worries for misusing `CircomConfig` was not that bad. Currently, the user can only provide a string that matches the gadget they want, however, with the `CircomGadget`/`CircomCoprocessor` interfaces, its hard to mess this up and smooth to implement.

I'm looking forward to any feedback about how the code looks!

huitseeker commented on Cargo.toml:
```diff
@@ -147,6 +148,3 @@ harness = false
 [[bench]]
 name = "public_params"
 harness = false
-
-[patch.crates-io]
-sppark = { git = "https://github.com/supranational/sppark", rev="5fea26f43cc5d12a77776c70815e7c722fd1f8a7" }
```
This makes our tests and benches bearable on M1/M2, see https://github.com/supranational/pasta-msm/issues/2

huitseeker commented on Cargo.toml:
```diff
@@ -53,6 +53,9 @@ serde_repr = "0.1.10"
 stable_deref_trait = "1.1.1"
 thiserror = { workspace = true }
+byteorder = "1.4.3"
+nova-scotia = { git = "https://github.com/huitseeker/Nova-Scotia.git", branch = "winston-main", features = ["circom-2"] }
```
I expect we'll need a better home there.

huitseeker commented on src/cli/paths.rs:
```diff
@@ -35,18 +32,26 @@ pub mod non_wasm {
+    }
+
+    pub fn circom_binary() -> PathBuf {
+        circom_gadgets().join("circom")
```
what's the idea behind adding the gadgets prefix?

winston-h-zhang replied:
I wanted to organize all circom related things under the `circom-gadgets` directory. Now that I type it out though, maybe `circom-gadgets` should be renamed to `circom`

huitseeker commented on src/coprocessor/circom.rs:
```diff
@@ -0,0 +1,115 @@
+};
+
+/// TODO: fix this duplication with the one in `cli::paths`
+fn circom_gadgets() -> PathBuf {
```
How would do you recommend we populate the circom binary in CI? Have it installed at the host? OK, but where?
Can I pass its location as an environment variable?

winston-h-zhang replied:
Yes, I guess there should be an environment variable for this, then the host will have to install it

huitseeker commented on src/cli/circom.rs:
```diff
@@ -0,0 +1,84 @@
+
+use super::paths::non_wasm::{circom_binary, circom_gadgets};
+
+/// Attempts to verify the checksum and version of the circom binary.
+/// If the binary does not exist, it is downloaded.
+/// TODO
```
This may help:
https://gist.github.com/huitseeker/ff00e7ca970b2c12f62e2d58542e9a1c
As most of our code base isn't async, `smol::block_on` may help as well:
https://github.com/smol-rs/smol

winston-h-zhang replied:
Thank you!

huitseeker reviewed (commented):
This is a great start! I think the first thing we want to get to, besides the details I added inline, is an integration test that runs through CI.

Feel free to modify the github actions we have for the moment adding a separate job that installs circom (only on Linux) and runs through this example.

winston-h-zhang commented:
So I tried looking at adding an integration test, but got stuck in looking at all the setup required. I'm mostly not sure where to put such a test in CircleCI.

huitseeker commented on examples/circom.rs:
```diff
@@ -0,0 +1,119 @@
+    SC(CircomCoprocessor<F, CircomSha256<F>>),
+}
+
+/// Run the example in this file with
+/// `cargo run --release --example circom`
```
Except I don't have a circom source at hand as part of this PR, do I? It sure would be nice to have the module-level documentation tell me what pre-requisite commands I need to issue for this example to work.

winston-h-zhang replied:
Done

huitseeker commented on src/coprocessor/circom.rs:
```diff
@@ -0,0 +1,159 @@
+Available circom gadgets:{avaliable}
+
+If you want to setup a new circom gadget `{name}`, run
+    `lurk coprocessor --name {name} <{}_FOLDER>`",
```
what should I expect to find in the NAME_FOLDER for this to be valid?

winston-h-zhang replied:
Expanded

huitseeker commented on src/coprocessor/circom.rs:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    impl<F: LurkField, C: CircomGadget<F>> CircomCoprocessor<F, C> {
+        /// Creates a CircomConfig by loading in the data in `.lurk/circom-gadgets/<name>/*`
```
Is this PR to be interpreted under the assumption some gadgets are available locally, but are not part of this PR?

huitseeker reviewed (commented):
I'm not sure I understand all the moving pieces this PR is supposed to operate with. Until we get to chat about it offline, let's see if I can give preliminary comments that help review:
1. rebase this on a fresher main (the CI triggers are outdated),
2. The immediate GH CI issue is that pkg-config, a helper tool that provides compile and link flags for libraries, is not found. In GH CI, you can take the beginning of the machine setup and add:
```
- name: Install dependencies
  run: sudo apt-get install -y pkg-config libssl-dev
```
(or, if this is self-hosted, add it to https://github.com/lurk-lab/gh-actions-runner Docker image)
- the reason the above happens is because reqwest uses the local OpenSSL instance : https://github.com/seanmonstar/reqwest#requirements
- it's unfortunate that the sha256 example is hollow, what would it take to complete it?

winston-h-zhang commented on src/cli/mod.rs:
```diff
@@ -340,6 +352,7 @@ impl ReplCli {
             &self.public_params_dir,
             &self.proofs_dir,
             &self.commits_dir,
+            &None, // TODO: help
```
@samuelburnham I ran into these changes when rebasing. For the sake of compiling, I put None here, but this is probably not correct. Should `circom` configuration files go with `LURK_DIR`s? I'm not sure what the final structures here will look like; what should be done?

samuelburnham replied:
As mentioned in DMs this could be `&self.circom_dir` and have the same behavior as the other `LURK_DIRS`

winston-h-zhang replied:
Done, hopefully this matches the behavior of the other directories, I've requested a review

winston-h-zhang commented:
@huitseeker I've added in the circom sha256 files into the `examples/` folder (which is blowing up the diff again, but it makes the user intro easier to manage). I've also added a bunch of documentation, let me know what you think.

Will have to talk to Sam and perhaps Arthur to figure out how we should manage all these lurk paths.

samuelburnham commented on src/cli/paths.rs:
```diff
@@ -95,11 +110,13 @@ pub(crate) fn set_lurk_dirs(
     );
     let proofs = get_path(proofs_dir, "proofs", proofs_default_dir());
     let commits = get_path(commits_dir, "commits", commits_default_dir());
+    let circom = get_path(circom_dir, "circom", commits_default_dir());
```
This should be `circom_default_dir`

samuelburnham commented on src/cli/paths.rs:
```diff
@@ -43,7 +44,11 @@ pub(crate) fn proofs_default_dir() -> Utf8PathBuf {
 
 pub(crate) fn commits_default_dir() -> Utf8PathBuf {
-    Utf8PathBuf::from("commits")
+    Utf8PathBuf::from(".lurk/commits")
```
What's the rationale for this change? Existing behavior is for `commits` to be located in the local workdir as commitments can contain private data and shouldn't be shared

samuelburnham replied:
Nvm I was mistaken, it should default to $HOME/.lurk/commits (see DMs)

samuelburnham commented on src/cli/paths.rs:
```diff
@@ -43,7 +44,11 @@ pub(crate) fn proofs_default_dir() -> Utf8PathBuf {
+    Utf8PathBuf::from(".lurk/commits")
+}
+
+pub(crate) fn circom_default_dir() -> Utf8PathBuf {
```
I believe you want the following:
```
#[cfg(not(target_arch = "wasm32"))]
pub(crate) fn circom_default_dir() -> Utf8PathBuf {
    let home = home::home_dir().unwrap();
    Utf8PathBuf::from_path_buf(home.join(".lurk/circom")).expect("path contains invalid Unicode")
}
#[cfg(target_arch = "wasm32")]
pub(crate) fn circom_default_dir() -> Utf8PathBuf {
    Utf8PathBuf::from(".lurk/circom")
}
```

samuelburnham commented on src/cli/mod.rs:
```diff
@@ -376,6 +389,18 @@ impl Cli {
             }
+            #[allow(unused_variables)]
+            Command::Circom(circom_args) => {
+                #[cfg(not(target_arch = "wasm32"))]
```
+1 on removing `non_wasm` modules where possible, as most of the CLI compiles to Wasm (the `home` crate being an exception)

winston-h-zhang replied:
It's been removed 🎉

samuelburnham commented on src/cli/circom.rs:
```diff
@@ -0,0 +1,124 @@
+    ///  1. `LURK_CIRCOM_PATH`
+    ///  2. `.lurk/circom/circom`
+    ///
+    /// We *do not* consider the case where the user already has some
+    /// `circom` binary downloaded. The user will have two possibly
+    /// conflicting circom binaries floating around. However, things
+    /// should be kept separate as Lurk will never touch the user binary
+    /// and the user should never manually call the Lurk Circom binary.
```
Perhaps this could be configurable so that the user isn't forced to have duplicate binaries? One idea would be to list the `circom` binary as a pre-requisite in the docs, search for the `CIRCOM` env var or its default location when running the CLI, and if it can't be found only then download the binary to `LURK_CIRCOM_PATH`, or simply return an error.

winston-h-zhang commented:
Ok, everything seems to be almost working, except for `reqwest::blocking` for wasm. @huitseeker is there a way to make this work, or do we have to avoid downloading circom on wasm?

arthurpaulino commented on src/cli/mod.rs:
```diff
@@ -90,6 +102,10 @@ struct LoadArgs {
     #[clap(long, value_parser)]
     commits_dir: Option<Utf8PathBuf>,
+
+    /// Path to commitments directory
+    #[clap(long, value_parser)]
+    circom_dir: Option<Utf8PathBuf>,
```
Typo in the docstring

arthurpaulino commented on src/cli/mod.rs:
```diff
@@ -43,6 +44,17 @@ enum Command {
     Repl(ReplArgs),
     /// Verifies a Lurk proof
     Verify(VerifyArgs),
+    /// Instansiates a new circom gadget to interface with bellperson.
+    ///
+    /// To setup a new circom gadget `<NAME>`, place your circom files in a designated folder and
+    /// create a file called `<NAME>.circom`. `<CIRCOM_FOLDER>/<NAME>.circom` is the input file
+    /// for the `circom` binary; in this file you must declare your circom main component.
+    ///
+    /// Then run `lurk circom --name <NAME> <CIRCOM_FOLDER>` to instansiate a new gadget `<NAME>`.
+    /// The new components are stored in `.lurk/circom/<NAME>/*`.
+    // (TODO: replace `.lurk` with env var)
+    #[command(verbatim_doc_comment)]
+    Circom(CircomArgs),
```
Unfortunately a text this big will pollute the help menu

huitseeker replied:
`help` can show help for other subcommands, see https://github.com/clap-rs/clap/issues/416

arthurpaulino replied:
I understand. What I'm saying is that this wall of text will take the entire space for `lurk help`. We better have a short version and then a long version for `lurk circom --help`

winston-h-zhang replied:
Will do

winston-h-zhang commented:
@huitseeker How should I deal with the licensing?

huitseeker commented on src/cli/circom.rs:
```diff
@@ -0,0 +1,118 @@
+    use std::io::Write;
+
+    let url = match std::env::consts::OS {
+        "linux" => "https://github.com/iden3/circom/releases/download/v2.1.6/circom-linux-amd64",
+        "macos" => "https://github.com/iden3/circom/releases/download/v2.1.6/circom-macos-amd64",
+        "windows" => {
+            "https://github.com/iden3/circom/releases/download/v2.1.6/circom-windows-amd64.exe"
```
Those URLs should use `CIRCOM_VERSION`

huitseeker commented on src/cli/circom.rs:
```diff
@@ -0,0 +1,118 @@
+            "https://github.com/iden3/circom/releases/download/v2.1.6/circom-windows-amd64.exe"
+        }
+        _ => {
+            eprintln!("Unsupported OS");
+            std::process::exit(1);
```
I would rather panic with an error message saying the Circom binary was not found *and* could not be downloaded. On an unsupported file system, it is OK for the user to provide this binary on their own.

Which makes me think this should probably modify the CircleCI config to install this binary on the Arm64 machines.

huitseeker replied:
You could even return an `Error`, as your caller would know how to handle it.

huitseeker commented on src/circuit/gadgets/circom/mod.rs:
```diff
@@ -0,0 +1,26 @@
+//! # Usage of circom coprocessors.
```
I would advise putting `#[deny(missing_docs)]` at the top of the module and fixing the errors.

huitseeker commented on src/cli/circom.rs:
```diff
@@ -0,0 +1,118 @@
+/// where `<CIRCOM_DIR>` can be configured via the config file,
+/// a environment variable, or through a CLI argument, in that order.
+///
+/// We *do not* consider the case where the user already has some
+/// `circom` binary downloaded. The user will have two possibly
+/// conflicting circom binaries floating around. However, things
+/// should be kept separate as Lurk will never touch the user binary
+/// and the user should never manually call the Lurk Circom binary.
```
```suggestion
/// We *do not* consider the case where the user already has some
/// `circom` binary available in his $PATH. The user will have two possibly
/// conflicting circom binaries floating around. However, things
/// should be kept separate as Lurk will never touch the user binary
/// and the user should never manually call the Lurk Circom binary.
```

huitseeker commented on src/cli/circom.rs:
```diff
@@ -0,0 +1,118 @@
+    };
+
+    if success {
+        Ok(Command::new(circom_path))
```
Nit: you’ve defined this object above, would it be possible to reuse this definition?

winston-h-zhang replied:
the `.arg().output()` consumes the first

huitseeker commented on src/coprocessor/circom.rs:
```diff
@@ -0,0 +1,164 @@
+        ) -> Result<(AllocatedPtr<F>, AllocatedPtr<F>, AllocatedContPtr<F>), SynthesisError>
+        {
+            let input = self.gadget.clone().into_circom_input(input_exprs);
+            let witness = circom_scotia::calculate_witness(&self.config, input, true).expect("msg");
```
I think we can do better than `.expect(“msg”)` specially for a function that returns `Result`

huitseeker commented on src/cli/mod.rs:
```diff
@@ -463,10 +516,24 @@ impl Cli {
+            Command::Circom(circom_args) => {
+                use crate::cli::circom::create_circom_gadget;
+                if circom_args.name == "main" {
+                    bail!("Circom gadget name cannot be `main`, for internal circom reasons")
```
```suggestion
                    bail!("Circom gadget name cannot be `main`, see circom documentation")
```

huitseeker commented on src/cli/circom.rs:
```diff
@@ -0,0 +1,118 @@
+    );
+    fs::create_dir_all(&circom_gadget)?;
+    let output = get_circom_binary()?
+        .arg(circom_file)
+        .arg("--r1cs")
+        .arg("--wasm")
+        .arg("--output")
+        .arg(&circom_gadget)
+        .arg("--prime")
+        .arg("vesta")
+        .output()
+        .expect("circom failed");
```
Are we sure the `vesta` prime is the one we should use? what about bn256 support?
In practice, the CLI has an idea of what field and proving system we’re being called with, and should pass that to you here.

Nit: You can put arg in the same line for concision `.args(&["--r1cs", "--wasm", "--output", "--prime", "vesta"])`

huitseeker reviewed (commented):
This is very close to the finish line! The comments marked ‘nit’ are stylistic and can safely be ignored.

winston-h-zhang commented:
@huitseeker I've addressed most of the changes; except I can't get the circom-integration CI test working. Is there any way to check out two repos and access files between them?

huitseeker commented on src/cli/circom.rs:
```diff
@@ -0,0 +1,132 @@
+            format!("https://github.com/iden3/circom/releases/download/v{CIRCOM_VERSION}/circom-windows-amd64.exe")
+        }
+        os => {
+            bail!("Unsupported OS: {os}. Unable to automatically download the necessary circom binary, please manually download Circom v2.1.6 to `.lurk/circom/circom`");
```
Good edit. This should use CIRCOM_VERSION as well, right?

huitseeker commented on examples/circom.rs:
````diff
@@ -0,0 +1,169 @@
+//! ```
+//! pragma circom 2.1.6;
+//!
+//! include "circomlib/sha256/sha256_2.circom";
````
I needed the following change:
```suggestion
//! include "./sha256/sha256_2.circom";
```

huitseeker commented on src/cli/mod.rs:
```diff
@@ -92,6 +98,10 @@ struct LoadArgs {
     #[clap(long, value_parser)]
     commits_dir: Option<Utf8PathBuf>,
+
+    /// Path to commitments directory
```
Nit: this comment is wrong

huitseeker commented on src/coprocessor/circom.rs:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    fn validate_gadget<F: LurkField, C: CircomGadget<F>>(gadget: &C) -> Result<()> {
+        if !circom_dir().exists() {
```
So the workflow suggested in the example (great doc comments BTW) fails, because it reaches this point, at which the setup of the `LURK_DIRS` variable has not been achieved.

Full logs:
https://gist.github.com/huitseeker/ec7ae2675b93f2ec383a601c4376d013

Indeed, we hit the expect in `circom_dir()` because we have not called the prior required function of `set_lurk_dirs()`, as the panic message indicates. Except it's hard to call into this initialization, since at this entry point you don't have access to the cli configuration. Without a default config, the best call you can make here is:
```rust
set_lurk_dirs(&HashMap::new(), &None,& None, &None, &None);
```
We should probably have a default configuration file that would make this better - in other terms, the path configuration done in the cli's path module should have two entry points:
- one through a library, as in the calling code here, which can use env vars, a configuration file found in a conventional location, and a sane hardcoded default,
- one through the CLI, which can use the CLI's arguments on top of the above.

and most of all, *that setup should be idempotent, encapsulated, and at no point should it include any public function that can fail because function calls have not been made in the right order*.

Overall, this is not the fault of this PR, but simply because since at least #581 patently **wrong** usage of the `LURK_DIRS` `OnceCell`. Instead of asking caller to keep a state machine in their head, the path setup instead use `OnceCell::get_or_init` and not rely on error messages and testing (which in this case is non-existent, since I'm apparently the first to run through this code on a fresh state 😠 ) to hope for the programmer to use the API correctly. `OnceCell`s are meant to obviate the need for complex sequencing of initialization operations, let's use them properly.

This PR should I think dump the trivial call to `set_lurk_dirs` above. This overrides the directory setup, but I don't know how to otherwise fix this mess in a reasonable time frame -- and I will open an issue for the later cleanup.

huitseeker commented on examples/circom.rs:
```diff
@@ -0,0 +1,169 @@
+
+impl<F: LurkField> CircomGadget<F> for CircomSha256<F> {
+    fn name(&self) -> &str {
+        "circom_sha25"
```
This is wrong for the intended example, you instead want `"sha256_2"`

huitseeker reviewed (changes requested):
I've left comments inline. If it helps, you can find a minimal gist that makes these examples run here:
https://gist.github.com/huitseeker/52989f2ca2a29a2f4284d8a7a3f15b35
But see also required changes not included here, such as [this comment](https://github.com/lurk-lab/lurk-rs/pull/555#discussion_r1298449908)

huitseeker commented on examples/circom.rs:
````diff
@@ -0,0 +1,159 @@
+//!
+//! Now return to the `lurk-rs` directory and run the following commands
+//! ```
+//! cargo run --release -- circom --name sha256_2_test <PATH_TO_CIRCOMLIB>/tests/circuits/
````
This path is wrong by one character, please fix.

huitseeker commented on src/coprocessor/circom.rs:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    fn validate_gadget<F: LurkField, C: CircomGadget<F>>(gadget: &C) -> Result<()> {
+        // TODO: This is a temporary hack, https://github.com/lurk-lab/lurk-rs/pull/555#discussion_r1298524025
```
Nit: you could link to https://github.com/lurk-lab/lurk-rs/issues/621 instead

huitseeker commented on src/coprocessor/circom.rs:
```diff
@@ -0,0 +1,180 @@
+        print_error(gadget.name(), subdirs)
+    }
+
+    /// A concrete instansiation of a [CircomGadget] with a corresponding [CircomConfig] as a coprocessor.
```
Nit: typo

huitseeker commented on src/cli/mod.rs:
```diff
@@ -45,6 +46,11 @@ enum Command {
     Repl(ReplArgs),
     /// Verifies a Lurk proof
     Verify(VerifyArgs),
+    /// Instansiates a new circom gadget to interface with bellperson.
```
Nit: "Instantiates"

huitseeker commented on src/cli/mod.rs:
```diff
@@ -450,6 +474,35 @@ struct VerifyArgs {
+/// create a file called `<NAME>.circom`. `<CIRCOM_FOLDER>/<NAME>.circom` is the input file
+/// for the `circom` binary; in this file you must declare your circom main component.
+///
+/// Then run `lurk circom --name <NAME> <CIRCOM_FOLDER>` to instansiate a new gadget `<NAME>`.
```
Nit: instantiate

huitseeker commented on src/coprocessor/circom.rs:
```diff
@@ -0,0 +1,180 @@
+create a file called `{name}.circom`. The circom binary expects `{}_FOLDER>/{name}.circom` 
+as the input file; in this file you must declare your circom main component.
+
+Then run `lurk coprocessor --name {name} <{}_FOLDER>` to instansiate a new gadget `{name}`.",
```
"Instantiate"

huitseeker reviewed (commented):
Thanks a lot, this looks great and essentially finished, except for :
- the path to the `sha256_2_test.circom` file,
- the custom branch of circom-scotia
See nits inline

winston-h-zhang commented:
🤦

huitseeker commented on Cargo.toml:
```diff
@@ -62,6 +62,11 @@ camino = { workspace = true }
 abomonation_derive = { git = "https://github.com/lurk-lab/abomonation_derive.git" }
 crossbeam = "0.8.2"
+byteorder = "1.4.3"
+circom-scotia = { git = "https://github.com/lurk-lab/circom-scotia", branch = "bump-bellpepper" }
```
Never mind, that's actually fine! Sorry for the brain fart.
BUT please look at https://github.com/lurk-lab/circom-scotia/pull/18 as you'll note there that 2/3 of the dependencies your PR uses are not needed

huitseeker reviewed (approved):
Tip top! Thank you so much for the thorough work on this one, @winston-h-zhang

emmorais commented on examples/circom.rs:
````diff
@@ -0,0 +1,159 @@
+//! }
+//! ```
+//!
+//! Hooray! Now we can use a [CircomSha256] coprocessor just like a normal one.
````
Happy to read that comment! Nice work @winston-h-zhang !
